GPU/Debugger/Record.cpp

```diff
@@ -0,0 +1,1013 @@
+			ibytes = vcount * sizeof(u8);
+		} else if (idx == GE_VTYPE_IDX_16BIT) {
+			ibytes = vcount * sizeof(u16);
+		} else if (idx == GE_VTYPE_IDX_32BIT) {
+			ibytes = vcount * sizeof(u32);
+		}
+	} else {
+		vbytes = vcount * vdec.VertexSize();
+	}
+}
+
+static const u8 *mymemmem(const u8 *haystack, size_t hlen, const u8 *needle, size_t nlen) {
+    if (!nlen) {
+        return nullptr;
+	}
+
+	const u8 *last_possible = haystack + hlen - nlen;
+    int first = *needle;
+    const u8 *p = haystack;
+    while (p <= last_possible) {
+		p = (const u8 *)memchr(p, first, last_possible - p + 1);
+		if (!p) {
+			return nullptr;
+		}
+        if (!memcmp(p, needle, nlen)) {
+            return p;
+		}
+
+        p++;
+    }
+
+    return nullptr;
+}
+
+static Command EmitCommandWithRAM(CommandType t, const void *p, u32 sz) {
+	FlushRegisters();
+
+	Command cmd{t, sz, 0};
+
+	if (sz) {
+		// If at all possible, try to find it already in the buffer.
+		const u8 *prev = nullptr;
+		const size_t NEAR_WINDOW = std::max((int)sz * 2, 1024 * 10);
+		// Let's try nearby first... it will often be nearby.
+		if (pushbuf.size() > NEAR_WINDOW) {
+			prev = mymemmem(pushbuf.data() + pushbuf.size() - NEAR_WINDOW, NEAR_WINDOW, (const u8 *)p, sz);
+		}
+		if (!prev) {
+			prev = mymemmem(pushbuf.data(), pushbuf.size(), (const u8 *)p, sz);
+		}
+
+		if (prev) {
+			cmd.ptr = (u32)(prev - pushbuf.data());
+		} else {
+			cmd.ptr = (u32)pushbuf.size();
+			int pad = 0;
+			if (cmd.ptr & 0xF) {
+				pad = 0x10 - (cmd.ptr & 0xF);
+				cmd.ptr += pad;
+			}
+			pushbuf.resize(pushbuf.size() + sz + pad);
+			if (pad) {
+				memset(pushbuf.data() + cmd.ptr - pad, 0, pad);
+			}
+			memcpy(pushbuf.data() + cmd.ptr, p, sz);
+		}
+	}
+
+	commands.push_back(cmd);
+
+	return cmd;
+}
+
+static void EmitTextureData(int level, u32 texaddr) {
+	GETextureFormat format = gstate.getTextureFormat();
+	int w = gstate.getTextureWidth(level);
+	int h = gstate.getTextureHeight(level);
+	int bufw = GetTextureBufw(level, texaddr, format);
+	int extraw = w > bufw ? w - bufw : 0;
+	u32 sizeInRAM = (textureBitsPerPixel[format] * (bufw * h + extraw)) / 8;
+
+	u32 bytes = Memory::ValidSize(texaddr, sizeInRAM);
+	if (Memory::IsValidAddress(texaddr)) {
+		FlushRegisters();
+
+		CommandType type = CommandType((int)CommandType::TEXTURE0 + level);
+		const u8 *p = Memory::GetPointerUnchecked(texaddr);
+
+		// Dumps are huge - let's try to find this already emitted.
+		for (u32 prevptr : lastTextures) {
+			if (pushbuf.size() < prevptr + bytes) {
+				continue;
+			}
+
+			if (memcmp(pushbuf.data() + prevptr, p, bytes) == 0) {
+				commands.push_back({type, bytes, prevptr});
+				// Okay, that was easy.  Bail out.
+				return;
+			}
+		}
+
+		// Not there, gotta emit anew.
+		Command cmd = EmitCommandWithRAM(type, p, bytes);
+		lastTextures.push_back(cmd.ptr);
+	}
+}
+
+static void FlushPrimState(int vcount) {
+	// TODO: Eventually, how do we handle texturing from framebuf/zbuf?
```

Comment:
Not right now.  The recording isn't aware of what regions of VRAM are textures copied into VRAM (before the frame but not included in the dump, or during the frame, etc.), and which are rendered sections.

For example, suppose I record a scene with bloom using the softgpu.  At the end, it will have the bloom texture in the dump file, and it will always render from it - regardless of what changes are made to the rendering that (in the game) generated the texture.

For 1:1 render-to-texture, this is potentially solvable - but it doesn't currently try to track framebufs or depthbufs.

-[Unknown]